tornettools/generate_tgen.py

```diff
@@ -232,6 +245,42 @@ def __convert_privcount_key_to_tgen_key(str):
     else:
         return str
 
+def __calculate_hiddeen_service_onion_url(pem_key):
+    # load key
+    key = RSA.importKey(pem_key)
+    if key.has_private():
+        key = key.publickey()
+
+    # calculate onion address
+    onion_address = hashlib.sha1(key.exportKey('DER')[22:]).digest()[:10]
+
+    # and b32 encode it
+    return base64.b32encode(onion_address).decode('utf-8').lower()+'.onion'
+
+def __generate_hidden_service_private_key():
+    # generate private/public key pair
+    key = rsa.generate_private_key(backend=default_backend(), public_exponent=65537, \
+        key_size=1024)
+
+    # get private key in PEM container format
+    pem = key.private_bytes(encoding=serialization.Encoding.PEM,
+        format=serialization.PrivateFormat.TraditionalOpenSSL,
+        encryption_algorithm=serialization.NoEncryption())
+
+    # decode to printable strings
+    private_key_str = pem.decode('utf-8')
+
+    return private_key_str
+
+def __get_hidden_service_privkey_and_url():
+    # generate RSA 1024 privkey
+    privkey = __generate_hidden_service_private_key()
+
+    # calculate onion url based on key
+    onion_url = __calculate_hiddeen_service_onion_url(privkey)
+
+    return (privkey, onion_url)
```

Comment:
I've always generated these with Tor itself, so I'm not sure if Tor will consider the keys created with python as valid. (Tor's log file in `shadow.data/hosts/server/*log` would probably include a warning if it is invalid.)

Reply:
It looks like it worked. Anyway, I've reworked this to support v3, as per Steven's suggestion.